gpu/preconditioner/block_jacobi_kernels.cu

```diff
@@ -166,6 +166,220 @@ GKO_ENABLE_IMPLEMENTATION_SELECTION(select_apply, apply);
 using compiled_kernels = syn::compile_int_list<1, 13, 16, 32>;
 
 
+template <typename IndexType>
+__global__ void find_natural_blocks_kernel1(size_type m, IndexType *row,
+                                            bool *v)
+{
+    size_type tidx = blockDim.x * blockIdx.x + threadIdx.x;
+
+    if (tidx >= m) {
+        return;
+    }
+    IndexType nnzthis = row[tidx + 1] - row[tidx];
+    IndexType nnzlast = (tidx > 0) ? (row[tidx] - row[tidx - 1]) : 0;
+    v[tidx] = (nnzthis == nnzlast) ? 0 : 1;
+}
+
+template <typename IndexType>
+__global__ void find_natural_blocks_kernel2(size_type m, const IndexType *row,
+                                            const IndexType *col, bool *v)
+{
+    size_type tidx = blockDim.x * blockIdx.x + threadIdx.x;
+    size_type i = threadIdx.x;
+    int32 lastcol = -1;
+    int32 thiscol = -1;
+    int32 different = 0;
+    bool loc = 0;
+
+    if (tidx >= m) {
+        return;
+    }
+    if (v[tidx] == 0) {
+        size_type bound = row[tidx + 1] - row[tidx];
+        for (size_type j = 0; j < (bound + 32) / 32; j++) {
+            size_type ii = i + 32 * j;
+            if (ii < bound) {
+                thiscol = col[row[tidx] + ii];
+                lastcol = col[row[tidx - 1] + ii];
+                if (thiscol != lastcol) {
+                    different = 1;
+                }
+            }
+        }
+        loc = __any(different == 1);
+        if (i == 0) {
+            v[tidx] = loc;
+        }
+    }
+}
+
+
+template <typename IndexType>
+__global__ void find_natural_blocks_kernel3(size_type num_rows,
+                                            int32 max_block_size, const bool *v,
+                                            IndexType *block_ptrs,
+                                            size_type *num_blocks_arr)
+{
+    size_type tidx = blockDim.x * blockIdx.x + threadIdx.x;
+    if (tidx >= 1) {
```

Comment:
So that means this kernel isn't parallelised yet right?

Reply:
Thank you for the explanation. Indeed I also thought parallelisation would be hard when looking at the code for block pointer creation. I'm not saying it has to be parallelised, I was just really wondering.

In this case, why do you generate 32 threads? Is that to ensure you do not have half warps (or worst) that other potential concurrent kernel would run on, when MS is activated? Or is it something else?

Reply:
There is no reason for the 32 threads. I could also just run one. But I guess it does not matter, given that one warp has to be launched anyway.

Reply:
It matters from the point of code readability.

Reply:
> what do you think, we win parallelism but get a different (potentially worse) pattern.

It could be interesting to expose "parallelism level" as a user option. Then you could choose what's more important for you. It also doesn't mean that the pattern will be worse, you could get lucky and get numerically better blocks.

Reply:
Is the parallelism level something I should add now or at a later stage? We may in the end move to a different block identification strategy overall...

Reply:
Later, block identification (and different strategies of it) is something that is useful independent of block-Jacobi, so we should move it into a new component anyway (we can add the option once we do it). Having it tightly coupled with block-Jacobi also makes testing difficult.

For now you can just implement the kernels so they can accept (and work with) that extra parameter, and hard-code it to 1 in the function that calls the kernels. This will make things easier later.

Reply:
OK. Well, adding this functionality is not a straight forward thing, depending on the parallelism you ask for, you need a different number of additional steps populating the carryover. Maybe I should not have brought up that idea... I will see what @yanjen tells me in terms of the performance benefits he observed.

Reply:
@hartwiganzt What part of the code do you mean by multi-step row-pointer?

Reply:
@yanjen you remember the code we converted a dense matrix to csr? There we needed to generate the row-pointer. The default was using a single thread handling this, but as far as I remember you had a version starting the row pointer in different locations, then an additional kernel populating the carryover. I do not exactly remember whether this was faster, and if so, how much...

Reply:
OK, thanks. That sounds very moderate - pretty much system noise. I guess we do not care about 3%.

Reply:
Yes, it does not seem worth the trouble.